### PR TITLE
terminal: Fix non regex search to actually be non regex

### LINKED ISSUE #4880: Terminal search is always in to regex mode
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Searching for "." in the Terminal finds empty characters where there is no output, see screenshot.
When resizing the dock with the search active this sometimes crashes Zed (no logs).
Screenshot 2023-03-22 at 10 26 45

### Environment

Zed: v0.77.3 (stable)
OS: macOS 13.2.1
Memory: 16 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

_No response_

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.
If you only need the most recent lines, you can run the `zed: open log` command palette action to see the last 1000.

_No response_

### PR DESCRIPTION
**Pull request by N8th8n8el**
Alacritty seems to support only regex search out of the box.
This PR just escapes all special regex chars to make non regex search work as expected.
Disclaimer: New to Rust.

Release Notes:

-Fixed [#4880](https://github.com/zed-industries/zed/issues/4880).

### CODE REVIEW

File: crates/terminal_view/src/terminal_view.rs [hunk 1]
```diff
@@ -418,21 +420,6 @@ impl TerminalView {
         .detach();
     }
 
-    pub fn find_matches(
```
Comment (N8th8n8el) [inline crates/terminal_view/src/terminal_view.rs:421]: This seems unused or am I missing something?

**Comment by N8th8n8el** [comment]
Could someone have a look at the failed CI runs? Either it is lying or I just don't understand where this second import is coming from, because it's not where the compiler says it is.

**Comment by SomeoneToIgnore** [comment]
While I also do not see the very same error when building, there's one definitely wrong thing: `std::collections::HashSet` usage per se — we're supposed to use `collections::HashSet` which is its own crate with an overidden hasher.

**Comment by SomeoneToIgnore** [comment]
I merged https://github.com/zed-industries/zed/pull/7481 and assume that rebasing on `main` would help with the compilation issues here.

**Comment by N8th8n8el** [comment]
> I merged #7481 and assume that rebasing on `main` would help with the compilation issues here.

It did, thanks.

>While I also do not see the very same error when building, there's one definitely wrong thing: std::collections::HashSet usage per se — we're supposed to use collections::HashSet which is its own crate with an overidden hasher.

Thanks for the info, but I decided to use a plain array instead of a HashSet after all. Both lookups are O(1) here and a simple test showed that searching an array of that size is faster than a lookup in the set. The code is also a bit simpler.

File: crates/terminal_view/src/terminal_view.rs (hunk 1 above)
Comment (SomeoneToIgnore) [reply crates/terminal_view/src/terminal_view.rs:439]: I think it's us who missed that after a bit gpui -> gpui2 rewrite, and you're right.

**Review by SomeoneToIgnore — APPROVED** [review]
> Both lookups are O(1) here and a simple test showed that searching an array of that size is faster than a lookup in the set. The code is also a bit simpler.

I would really disagree with the `Both lookups are O(1) here` part but totally agree with the rest of the sentence.

Thank you for paying attention to all that and fixing!

**Comment by N8th8n8el** [comment]
> I would really disagree with the `Both lookups are O(1) here` part

The array of special chars has a fixed length (independent of input, aka `n`), so a lookup is O(1), regardless of the data structure. If the size of the data structure were to depend on the input it would be O(n) for the array and O(1) for the map.
But maybe you meant all the lookups combined (which is O(n) for both - one lookup of O(1) for each input char).
@SomeoneToIgnore

**Comment by SomeoneToIgnore** [comment]
Thank you for the elaboration, re-read it today with a fresh head — I think I was plain wrong when posted my original comment.

**Pull request merged**